soc/arm/microchip_mec/mec172x/soc.c

```diff
@@ -0,0 +1,417 @@
+	((struct vbatr_regs *)(DT_REG_ADDR(DT_NODELABEL(vbr))))
+
+#define CLK32K_SIL_OSC_DELAY		256U
+#define CLK32K_PLL_LOCK_WAIT		(16U * 1024U)
+#define CLK32K_PIN_WAIT			4096U
+#define CLK32K_XTAL_WAIT		4096U
+#define CLK32K_XTAL_MON_WAIT		(64U * 1024U)
+
+#define DEST_PLL			0
+#define DEST_PERIPH			1
+
+#define CLK32K_FLAG_CRYSTAL_SE		BIT(0)
+#define CLK32K_FLAG_PIN_FB_CRYSTAL	BIT(1)
+
+#define PIN_32KHZ_IN_MODE (MCHP_GPIO_CTRL_PUD_NONE |\
+	MCHP_GPIO_CTRL_PWRG_VTR_IO | MCHP_GPIO_CTRL_IDET_DISABLE |\
+	MCHP_GPIO_CTRL_MUX_F1)
+
+enum clk32k_src {
+	CLK32K_SRC_SIL_OSC = 0,
+	CLK32K_SRC_CRYSTAL,
+	CLK32K_SRC_PIN,
+	CLK32K_SRC_MAX
+};
+
+enum clk32k_dest { CLK32K_DEST_PLL = 0, CLK32K_DEST_PERIPH, CLK32K_DEST_MAX };
+
+static uint32_t spin_delay(uint32_t cnt)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+	uint32_t n;
+
+	for (n = 0U; n < cnt; n++) {
+		pcr->OSC_ID = n;
+	}
+
+	return n;
+}
+
+/*
+ * Make sure PCR sleep enables are clear except for crypto
+ * which do not have internal clock gating and UARTs.
+ */
+static int soc_pcr_init(void)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+
+	pcr->SYS_SLP_CTRL = 0U;
+
+	for (int i = 0; i < MCHP_MAX_PCR_SCR_REGS; i++) {
+		pcr->SLP_EN[i] = 0U;
+	}
+
+	pcr->SLP_EN[3] = MCHP_PCR3_CRYPTO_MASK;
+
+	return 0;
+}
+
+/*
+ * Initialize MEC172x EC Interrupt Aggregator (ECIA) and external NVIC
+ * inputs.
+ */
+static int soc_ecia_init(void)
+{
+	struct ecia_ar_regs *eciar = ECIA_XEC_REG_BASE;
+	struct ecs_regs *ecsr = ECS_XEC_REG_BASE;
+	uint32_t n;
+
+	mchp_pcr_periph_slp_ctrl(PCR_ECIA, MCHP_PCR_SLEEP_DIS);
+
+	ecsr->INTR_CTRL |= MCHP_ECS_ICTRL_DIRECT_EN;
+
+	/* gate off all aggregated outputs */
+	eciar->BLK_EN_CLR = 0xFFFFFFFFul;
+	/* gate on GIRQ's that are aggregated only */
+	eciar->BLK_EN_SET = MCHP_ECIA_AGGR_BITMAP;
+
+	/* Clear all GIRQn source enables */
+	for (n = 0; n < MCHP_GIRQ_IDX_MAX; n++) {
+		eciar->GIRQ[n].EN_CLR = 0xFFFFFFFFul;
+	}
+
+	/* Clear all external NVIC enables and pending status */
+	for (n = 0u; n < MCHP_NUM_NVIC_REGS; n++) {
+		NVIC->ICER[n] = 0xFFFFFFFFul;
+		NVIC->ICPR[n] = 0xFFFFFFFFul;
+	}
+
+	return 0;
+}
+
+static bool is_sil_osc_enabled(void)
+{
+	struct vbatr_regs *vbr = VBATR_XEC_REG_BASE;
+
+	if (vbr->CLK32_SRC & MCHP_VBATR_CS_SO_EN) {
+		return true;
+	}
+
+	return false;
+}
+
+static void enable_sil_osc(uint32_t delay_cnt)
+{
+	struct vbatr_regs *vbr = VBATR_XEC_REG_BASE;
+
+	vbr->CLK32_SRC |= MCHP_VBATR_CS_SO_EN;
+	spin_delay(delay_cnt);
+}
+
+static int enable_32k_pin(uint32_t wait_cnt)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+	struct gpio_ctrl_regs *gpcr = (struct gpio_ctrl_regs *)GPIO_BASE;
+
+	gpcr->CTRL_0165 = PIN_32KHZ_IN_MODE;
+
+	do {
+		if (pcr->PWR_RST_STS & MCHP_PCR_PRS_32K_ACTIVE_RO) {
+			pcr->CLK32K_SRC_VTR = MCHP_PCR_VTR_32K_SRC_SILOSC;
+			return 0;
+		}
+	} while (wait_cnt--);
+
+	return -ETIMEDOUT;
+}
+
+
+/* Do nothing if crystal configuration matches request. */
+static int enable_32k_crystal(uint32_t flags)
```

Comment:
this function always returns 0. Is that intended? Why return anything?

Reply:
Hi @sjg20 .  The caller is using the return value.  I did notice an issue in the routine and fixed that. The routine is only called if the build has been configured to use an external 32KHz clock source on a defined pin. The PCR HW can detect the pin has edges on it. The spin count is only long enough for HW to detect a few edges and set the HW status.